File: scratch/deploy.js
const { Client } = require('ssh2');
const SftpClient = require('ssh2-sftp-client');
const path = require('path');

const config = {
  host: '65.20.85.75',
  port: 22,
  username: 'graceandforce',
  password: 'wvpi2!ZnTcV];ncy',
  readyTimeout: 20000
};

const backendDir = path.resolve(__dirname, '../backend');
const setupScript = path.resolve(__dirname, '../vultr-setup.sh');
const nginxConf = path.resolve(__dirname, '../nginx-graceandforce.conf');

async function deploy() {
  const sftp = new SftpClient();
  const ssh = new Client();

  try {
    console.log('Connecting via SFTP...');
    await sftp.connect(config);
    
    console.log('Creating directories...');
    await executeSsh(ssh, config, 'sudo mkdir -p /var/www/grace-and-force/backend && sudo chown -R graceandforce:graceandforce /var/www');

    console.log('Uploading backend folder...');
    await sftp.uploadDir(backendDir, '/var/www/grace-and-force/backend');
    
    console.log('Uploading setup scripts...');
    await sftp.fastPut(setupScript, '/var/www/grace-and-force/vultr-setup.sh');
    await sftp.fastPut(nginxConf, '/var/www/grace-and-force/nginx-graceandforce.conf');

    console.log('Executing setup script...');
    // We pipe the password to sudo -S
    const setupCmd = `echo "${config.password}" | sudo -S bash /var/www/grace-and-force/vultr-setup.sh`;
    await executeSsh(ssh, config, setupCmd, true);

    console.log('Setting up Nginx...');
    const nginxCmd = `
      echo "${config.password}" | sudo -S cp /var/www/grace-and-force/nginx-graceandforce.conf /etc/nginx/sites-available/graceandforce
      echo "${config.password}" | sudo -S ln -sf /etc/nginx/sites-available/graceandforce /etc/nginx/sites-enabled/
      echo "${config.password}" | sudo -S nginx -t
      echo "${config.password}" | sudo -S systemctl reload nginx
    `;
    await executeSsh(ssh, config, nginxCmd, true);

    console.log('=== DEPLOYMENT COMPLETED SUCCESSFULLY ===');
  } catch (err) {
    console.error('Deployment Error:', err);
  } finally {
    await sftp.end();
  }
}

function executeSsh(sshClient, sshConfig, command, printOutput = false) {
  return new Promise((resolve, reject) => {
    // If not already connected
    sshClient.on('ready', () => {
      sshClient.exec(command, { pty: true }, (err, stream) => {
        if (err) return reject(err);
        
        let out = '';
        stream.on('close', (code, signal) => {
          sshClient.end();
          if (code !== 0) reject(new Error(`Command failed with code ${code}. Output: ${out}`));
          else resolve(out);
        }).on('data', (data) => {
          out += data.toString();
          if (printOutput) process.stdout.write(data.toString());
        }).stderr.on('data', (data) => {
          out += data.toString();
          if (printOutput) process.stderr.write(data.toString());
        });
      });
    }).on('error', reject).connect(sshConfig);
  });
}

deploy();

File: scratch/deploy.py
import paramiko
import sys

def deploy():
    host = "65.20.85.75"
    port = 22
    username = "graceandforce"
    password = "wvpi2!ZnTcV];ncy"

    print("Connecting to Vultr server via SSH...")
    client = paramiko.SSHClient()
    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    
    try:
        client.connect(host, port, username, password, timeout=10)
        print("Connected successfully!")
        
        script = """
set -e

export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"

echo "=== Finalizing Deployment ==="
cd ~/debate-engine/backend

echo ">>> Installing Python dependencies..."
pip3 install --break-system-packages --quiet psycopg2-binary python-dotenv google-generativeai httpx

echo ">>> Starting Server with PM2..."
npx pm2 restart grace-api || npx pm2 start server.js --name grace-api
npx pm2 save
echo "=== USER DEPLOYMENT COMPLETE ==="
"""

        print("Executing deployment script...")
        stdin, stdout, stderr = client.exec_command(script)
        
        for line in iter(stdout.readline, ""):
            print(line.encode('utf-8', 'ignore').decode('utf-8'), end="")
            
        for line in iter(stderr.readline, ""):
            print("ERROR: " + line.encode('utf-8', 'ignore').decode('utf-8'), end="")

        exit_status = stdout.channel.recv_exit_status()
        print(f"\\nCommand finished with exit status: {exit_status}")
        
    except Exception as e:
        print(f"Failed to connect or execute: {str(e)}")
    finally:
        client.close()

if __name__ == "__main__":
    deploy()
